File: java/org.chabu.test.node/src/org/chabu/nwtest/XferItem.java
package org.chabu.nwtest;

import java.util.Arrays;

import javax.xml.bind.annotation.XmlElement;
import javax.xml.bind.annotation.XmlElementWrapper;
import javax.xml.bind.annotation.XmlRootElement;
import javax.xml.bind.annotation.XmlTransient;

@XmlRootElement(name="XferItem")
public class XferItem {
	
	public enum Category{
		REQ, RES, EVT;
	}
	
	@XmlElement(name="Category")
	private Category category = Category.EVT;
	@XmlElement(name="Name")
	private String name = "";
	@XmlElement(name="CallIndex")
	private int callIndex = 0;

	@XmlElementWrapper(name="Parameters")
	@XmlElement(name="Parameter")
	private Parameter[] parameters = new Parameter[0];
	
	public XferItem(){
		
	}
	
	public void setCategory(Category cat) {
		this.category = cat;
	}

	@XmlTransient
	public Category getCategory() {
		return category;
	}
	public void setName(String name) {
		this.name = name;
	}
	@XmlTransient
	public String getName() {
		return name;
	}
	
	public void setCallIndex(int callIndex) {
		this.callIndex = callIndex;
	}
	@XmlTransient
	public int getCallIndex() {
		return callIndex;
	}
	public void setParameters(Parameter[] parameters) {
		this.parameters = parameters;
	}
	@XmlTransient
	public Parameter[] getParameters() {
		return parameters;
	}

	public String getValueString(String path) {
		ParameterValue pv = findParameterValue( path );
		return pv.value;
	}
	public long getValueLong(String path) {
		ParameterValue pv = findParameterValue( path );
		return Long.parseLong(pv.value);
	}
	public int getValueInt(String path) {
		ParameterValue pv = findParameterValue( path );
		return Integer.parseInt(pv.value);
	}
	public double getValueDouble(String path) {
		ParameterValue pv = findParameterValue( path );
		return Double.parseDouble(pv.value);
	}

	private ParameterValue findParameterValue(String path) {
		String[] parts = path.split("/");
		Parameter p = findParameterValue(parameters, parts, 0);
		if( p instanceof ParameterValue ){
			return (ParameterValue)p;
		}
		throw new RuntimeException("Not found: "+path);
	}

	private Parameter findParameterValue(Parameter[] par, String[] parts, int i) {
		String pathPart = parts[i];
		for( Parameter p : par ){
			if( pathPart.equals(p.getName())){
				if( p instanceof ParameterWithChilds ){
					ParameterWithChilds pc = (ParameterWithChilds)p;
					return findParameterValue(pc.getChilds(), parts, i+1);
				}
				return p;
			}
		}
		return null;
	}

	public void addParameter(String name, String value) {
		parameters = Arrays.copyOf(parameters, parameters.length+1);
		parameters[ parameters.length - 1 ] = new ParameterValue( name, value );
	}
	public void addParameter(String name, long value) {
		parameters = Arrays.copyOf(parameters, parameters.length+1);
		parameters[ parameters.length - 1 ] = new ParameterValue( name, value );
	}
	public void addParameter(String name, double value) {
		parameters = Arrays.copyOf(parameters, parameters.length+1);
		parameters[ parameters.length - 1 ] = new ParameterValue( name, value );
	}
	public void addParameter(String name, ParameterWithChilds value ) {
		parameters = Arrays.copyOf(parameters, parameters.length+1);
		parameters[ parameters.length - 1 ] = value;
	}

}

File: c/org.chabu.node/src/testprot/XferItem.cpp
/*
 * XferItem.cpp
 *
 *  Created on: 17.07.2016
 *      Author: Frank
 */

#include "XferItem.h"
#include <boost/format.hpp>
#include <boost/lambda/lambda.hpp>
#include <boost/lexical_cast.hpp>
#include <boost/algorithm/string.hpp>

#include "ParameterValue.h"
#include "ParameterWithChilds.h"

namespace testprot {

using std::string;
using std::string;

XferItem::XferItem() {
}

XferItem::~XferItem() {
}

void XferItem::load( pugi::xml_node node ){
	this->name      = node.child_value("Name");
	this->callIndex = std::atoi( node.child_value("CallIndex") );
	string enumName = node.child_value("Category");
	if( enumName == "REQ") this->category = Category::REQ;
	if( enumName == "RES") this->category = Category::RES;
	if( enumName == "EVT") this->category = Category::EVT;
	auto params = node.child("Parameters");
	if( params ){
		for (auto param = params.child("Parameter"); param; param = param.next_sibling("Parameter")){
			parameters.push_back( createParam(param));
		}
	}
}

void XferItem::encodeInto( pugi::xml_node node ){
	auto item = node.append_child("XferItem");
	item.append_child("Name").text().set(name.c_str());
	item.append_attribute("xmlns:xsi").set_value("http://www.w3.org/2001/XMLSchema-instance");
	item.append_attribute("xmlns:xsd").set_value("http://www.w3.org/2001/XMLSchema");

	{
		auto callIndexStr = boost::lexical_cast<std::string>(callIndex);
		item.append_child("CallIndex").text().set( callIndexStr.c_str() );
	}

	switch( this->category ){
	case Category::REQ:
		item.append_child("Category").text().set("REQ");
		break;
	case Category::RES:
		item.append_child("Category").text().set("RES");
		break;
	case Category::EVT:
		item.append_child("Category").text().set("EVT");
		break;
	}

	auto params = item.append_child("Parameters");
	for( auto it = parameters.begin(); it != parameters.end(); it++ ){
		auto p = *it;
		p->encodeInto(params);
	}
}

string XferItem::toString(){
	string parametersStr = joinParameterListToString( parameters, ",\n" );
	string categoryStr;
	switch( category ){
	case Category::REQ: categoryStr = "REQ"; break;
	case Category::RES: categoryStr = "RES"; break;
	case Category::EVT: categoryStr = "EVT"; break;
	}
	auto fmt = boost::format("XferItem{name=%s callIndex=%s category=%s parameters={%s}}") % name % callIndex % categoryStr % parametersStr;
	return fmt.str();
}


std::string XferItem::getValueString(std::string path) {
	ParameterValue pv = findParameterValue( path );
	return pv.value;
}
int64_t XferItem::getValueLong(std::string path) {
	ParameterValue pv = findParameterValue( path );
	return boost::lexical_cast<int64_t>(pv.value);
}
int32_t XferItem::getValueInt(std::string path) {
	ParameterValue pv = findParameterValue( path );
	return boost::lexical_cast<int32_t>(pv.value);
}
double XferItem::getValueDouble(std::string path) {
	ParameterValue pv = findParameterValue( path );
	return boost::lexical_cast<double>(pv.value);
}

ParameterValue& XferItem::findParameterValue(std::string path) {
	std::vector<std::string> parts;
	boost::split( parts, path, boost::lambda::_1 == '/');
	Parameter* p = findParameterValue(parameters, parts, 0);
	if( p && p->isParameterValue()){
		return p->asParameterValue();
	}
	throw std::runtime_error("Not found: "+path);
}
Parameter* XferItem::findParameterValue(std::vector<std::shared_ptr<Parameter>> par, std::vector<std::string> parts, int i) {
	std::string pathPart = parts[i];
	for( auto it = par.begin(); it != par.end(); it++ ){
		Parameter& p = **it;
		if( pathPart == p.name){
			if( p.isParameterWithChilds() ){
				ParameterWithChilds& pc = p.asParameterWithChilds();
				return findParameterValue(pc.childs, parts, i+1);
			}
			return &p;
		}
	}
	return nullptr;
}

void XferItem::addParameter(std::string name, std::string value) {
	auto p = std::shared_ptr<Parameter>{ new ParameterValue( name, value ) };
	parameters.push_back( p );
}
void XferItem::addParameter(std::string name, int64_t value) {
	auto p = std::shared_ptr<Parameter>{ new ParameterValue( name, boost::lexical_cast<std::string>(value) ) };
	parameters.push_back( p );
}
void XferItem::addParameter(std::string name, double value) {
	auto p = std::shared_ptr<Parameter>{ new ParameterValue( name, boost::lexical_cast<std::string>(value) ) };
	parameters.push_back( p );
}
void XferItem::addParameter( std::shared_ptr<Parameter> child) {
	parameters.push_back( child );
}
void XferItem::addParameters(std::vector<std::shared_ptr<Parameter>> parameters){
	for( auto it = parameters.begin(); it != parameters.end(); it++ ){
		this->parameters.push_back( *it );
	}
}

} /* namespace testprot */
